### PR TITLE
8346836: C2: Verify CastII/CastLL bounds at runtime

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch adds a develop flag `VerifyConstraintCasts`, which will verify the correctness of `CastIINode`s and `CastLLNode`s at runtime and crash the VM if the dynamic value lies outside the type value range.

Please take a look, thanks a lot.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8346836](https://bugs.openjdk.org/browse/JDK-8346836): C2: Verify CastII/CastLL bounds at runtime (**Enhancement** - P4)

### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**) Review applies to [8d140fd9](https://git.openjdk.org/jdk/pull/22880/files/8d140fd9d462aa87f837d24c7c7921bd5e4496ea)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Contributors
 * Vladimir Ivanov `<vlivanov@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22880/head:pull/22880` \
`$ git checkout pull/22880`

Update a local copy of the PR: \
`$ git checkout pull/22880` \
`$ git pull https://git.openjdk.org/jdk.git pull/22880/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22880`

View PR using the GUI difftool: \
`$ git pr show -t 22880`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22880.diff">https://git.openjdk.org/jdk/pull/22880.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22880#issuecomment-2561920497)
</details>

### CODE REVIEW

**Comment by merykitty** [comment]
Running tier 1 tests with this flag and `-XX:+StressGCM` reveals several failures. One example is `compiler.arraycopy.TestArrayCopyConjoint`. In the method `testByte(int, int, int)`, we have the final graph before matching:

image

`228 CastII` is a cast on `P0` has its type being `int[0, 32]`, it depends on 3 dominating ifs:
- `100 If`, which corresponds to `(P0 < 0) == false`
- `223 If`, which corresponds to `(P0 != 0) == true`
- `200 If`, which corresponds to `(ConvI2L(P0) u<= 32L) == true`

As a result, `228 CastII` should have its control flow being the `IfTrueNode` projection of `200 If`. However, it is incorrectly wired to the `IfTrueNode` projection of `223 If`, which leads to the verification failure.

**Review by eme64 — CHANGES_REQUESTED** [review]
Looks interesting :)

Would we have caught any bug with this, do you have an example?

If I remember correctly, we relax/widen the Cast ranges somewhere later in optimizations, so that different CastII etc can common. Probably happens after loop-opts. So the ranges usually go from `[1..10]` -> `[0, max]` or `[-1 .. 1]` -> `int`. So this verification would then not be super effective, right? Things might have gone wrong much earlier with bad assumptions. I mean it could still catch issues, but I'm not sure how likely that is?

TLDR: I'd like some more context / motivation for this patch ;)

And: you should have at least one plain test where you enable the flag, and it compiles everything required to run an empty `main` function.

**Comment by eme64** [comment]
Ah, I actually see that you have some examples. So you plan on introducing this flag first, and only then fixing the issues? But does it fail with a simple `java --version`? Or an empty `main` method, maybe with `-Xcomp`?

**Comment by merykitty** [comment]
@eme64 Thanks for looking at this

The context is that while reviewing #22666 I came to the conclusion that our handling of `depends_only_on_test` is broken. I have added a comment explaining my understanding and concerns there. In principle, before the execution, a `DivINode` is the same as a `CastIINode` which limits the value range of the divisor to `!= 0`. As a result, there should not be any difference in the way we handle the movements of these nodes. This leads me to the conclusion that `CastIINode`s may also be wired to the wrong control input, the reason we have not caught them is that unlike a division complaining loudly, a `CastIINode` will silently accept incorrect input values. This motivates me to make this patch.

> If I remember correctly, we relax/widen the Cast ranges somewhere later in optimizations, so that different CastII etc can common. Probably happens after loop-opts. So the ranges usually go from `[1..10]` -> `[0, max]` or `[-1 .. 1]` -> `int`.

You are right, it is in `ConstraintCastNode::widen_type` for which I will disable that widening in the presence of `VerifyConstraintCasts`.

> So you plan on introducing this flag first, and only then fixing the issues?

There are several failures in tier 1 alone, and this flag is not enabled by default or in the pipeline, so I think incorporating it first would be preferable, then after fixing all the issues we can add it to the stress options.

> But does it fail with a simple java --version? Or an empty main method, maybe with -Xcomp?

No it does not fail with `--version` or with an empty main method with and without `-Xcomp`.

File: src/hotspot/share/opto/castnode.cpp [hunk 1]
```diff
@@ -506,6 +506,12 @@ const Type* ConstraintCastNode::widen_type(const PhaseGVN* phase, const Type* re
   if (!phase->C->post_loop_opts_phase()) {
     return res;
   }
+
+  // Keep these casts for verification
+  if (VerifyConstraintCasts) {
```
Comment (eme64) [inline src/hotspot/share/opto/castnode.cpp:511]: Can we have 2 modes, one where we do the widening, and one where we do not?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 2]
```diff
@@ -837,6 +837,42 @@ void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register reg_rax,
...
+#ifdef ASSERT
+void C2_MacroAssembler::checked_cast_int(const TypeInt* type, Register dst) {
+  Label fail;
+  Label succeed;
+  cmpl(dst, type->_lo);
+  jccb(Assembler::less, fail);
+  cmpl(dst, type->_hi);
+  jccb(Assembler::lessEqual, succeed);
+  bind(fail);
+  stop("Invalid CastII");
```
Comment (eme64) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:849]: Is there any way to print the `[lo,hi]` and `dst`? I suppose one can always get that via the debugger, would just be convenient.

**Review by eme64 — COMMENTED** [review]
I like this idea a lot. I'm personally on the fence if it is ok to integrate a flag that does not yet pass in the tests. The risk is that nobody fixes it.

What do you think about that @TobiHartmann @vnkozlov ?

Did you run testing with and without the flag? Cannot see the link posted on JIRA.

**Comment by eme64** [comment]
Talked with @TobiHartmann , he thinks it is ok to integrate the flags, even if it has failures. Let's go ahead with it then.

It would still be nice to have 2 modes: one where we do the widening, and one without. Both may catch different things, right? I imagine it like this: `VerifyConstraintCasts=1` -> do widening. `VerifyConstraintCasts=2` -> disable widening.

Plus, can you do this?
> And: you should have at least one plain test where you enable the flag, and it compiles everything required to run an empty main function.

Or does that already trigger failures? It would be nice just so we have some basic testing, and see that the flag is not completely broken.

**Comment by eme64** [comment]
What about implementing the same for aarch64? That would increase our coverage eventually.

**Comment by vnkozlov** [comment]
We can add this flag to our stress testing sets of flags to make sure we run with it during our regular testing.

File: src/hotspot/share/opto/castnode.cpp (hunk 1 above)
Comment (merykitty) [reply src/hotspot/share/opto/castnode.cpp]: Done

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 2 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: Done, calling into VM somehow results in no stacktrace so I just call `hlt` instead.

**Comment by merykitty** [comment]
@eme64 Thanks for your reviews, I have added 2 test cases for `TestIterativeGVN` that set `VerifyConstraintCasts`, the name of the test may need to change but I have not been able to come up with anything preferable.

For aarch64, I don't have an aarch64 machine around so it would be not so trivial.

File: src/hotspot/share/opto/c2_globals.hpp [hunk 3]
```diff
@@ -660,6 +660,16 @@
...
+  develop(uint, VerifyConstraintCasts, 0,                                   \
+          "Perform runtime checks to verify the value of a "                \
+          "ConstraintCast lies inside its type"                             \
+          "0=does not perform any verification, "                           \
+          "1=perform verification on ConstraintCastNodes that are present " \
+          "during code emission, "                                          \
+          "2=Do not do widening of ConstraintCastNodes so that we can "     \
+          "have more verification coverage")                                \
+          range(0, 2)                                                       \
+                                                                            \
```
Comment (eme64) [inline src/hotspot/share/opto/c2_globals.hpp:672]: Nit: I think the alignment could be a little improved, maybe also the comments.
Here a quick draft:

```suggestion
  develop(uint, VerifyConstraintCasts, 0,                                   \
          "Perform runtime checks to verify the value of a ConstraintCast " \
          "lies inside its type"                                            \
          "0: No verification."                                             \
          "1: Verify (types are widened, same as without verification)."  \
          "2: Verify, where types are not widened, for better  " \
          "   verification coverage"). \
          range(0, 2)                                                       \
                                                                            \
```
I'm too lazy to fix the alignment of the `\`. This here is just a suggestion, I leave it up to you in the end ;)

File: src/hotspot/share/opto/castnode.cpp [hunk 4]
```diff
@@ -524,6 +524,12 @@ const Type* ConstraintCastNode::widen_type(const PhaseGVN* phase, const Type* re
   if (!phase->C->post_loop_opts_phase()) {
     return res;
   }
+
+  // Keep these casts for verification
```
Comment (eme64) [inline src/hotspot/share/opto/castnode.cpp:528]: You could write about `VerifyConstraintCasts` with its 2 modes, and say why they are both useful.

File: test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * @run main/othervm/timeout=300 -Xbatch -Xcomp -XX:+StressGCM -XX:VerifyConstraintCasts=1 compiler.c2.TestVerifyConstraintCasts
+ */
+
+/*
+ * @test
+ * @bug 8346836
+ * @requires vm.debug == true & vm.flavor == "server"
+ * @summary Run with -Xcomp to test -XX:+StressGCM -XX:VerifyConstraintCasts=2 in debug builds.
+ *
+ * @run main/othervm/timeout=300 -Xbatch -Xcomp -XX:+StressGCM -XX:VerifyConstraintCasts=2 compiler.c2.TestVerifyConstraintCasts
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java:39]:
```suggestion
 * @run main/othervm/timeout=300 -Xbatch -Xcomp -XX:VerifyConstraintCasts=2 compiler.c2.TestVerifyConstraintCasts
```
I would have left the `StressGCM` out - we already add that in our CI testing for some runs anyway. But I suppose it does not super hurt to have it in either.

Comment (eme64) [reply test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java:39]: What was your reason to put it in?

**Review by eme64 — CHANGES_REQUESTED** [review]
Just a few more comments :)

File: src/hotspot/share/opto/c2_globals.hpp (hunk 3 above)
Comment (merykitty) [reply src/hotspot/share/opto/c2_globals.hpp]: Done, I believe the common theme in this file is to use `=` to signify the meanings of different values of the same variable while `:` is to signify the meanings of different variables.

E.g:

      develop(uint, VerifyIterativeGVN, 0,                                      \
              "Verify Iterative Global Value Numbering"                         \
              "=XY, with Y: verify Def-Use modifications during IGVN"           \
              "          X: verify that type(n) == n->Value() after IGVN"       \
              "X and Y in 0=off; 1=on")                                         \
              constraint(VerifyIterativeGVNConstraintFunc, AtParse)             \

File: src/hotspot/share/opto/castnode.cpp (hunk 4 above)
Comment (merykitty) [reply src/hotspot/share/opto/castnode.cpp]: Done

File: test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java (hunk 5 above)
Comment (merykitty) [reply test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java:39]: Even if a `ConstraintCastNode` has a wrong control input, it is normally scheduled later than the control input. Using `-XX:+StressGCM` helps lift the node in the dominator tree, increases the chance a wrong value is successfully observed.

**Review by eme64 — APPROVED** [review]
Sorry for the delay.

Looks great, especially with the better comments! :clap:

**Review by vnkozlov — COMMENTED** [review]
Can we add AArch64 implementation too to cover our platforms?

**Comment by merykitty** [comment]
@vnkozlov I don't have an AArch64 machine so I feel less confident writing one. We can add an AArch64 implementation later, though. What do you think?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 6]
```diff
@@ -836,6 +836,52 @@ void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register reg_rax,
   // C2 uses the value of ZF to determine the continuation.
 }
 
+#ifdef ASSERT
+void C2_MacroAssembler::checked_cast_int(const TypeInt* type, Register dst) {
```
Comment (iwanowww) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:840]: Naming is a bit confusing here. It is a register which holds the value being range checked, not a register where new value is put.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 7]
```diff
@@ -836,6 +836,52 @@ void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register reg_rax,
...
+  BLOCK_COMMENT("CastII {");
+  Label fail;
+  Label succeed;
+  cmpl(dst, type->_lo);
```
Comment (iwanowww) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:844]: Optimization idea: some range checks may be redundant (when `lo`/`hi` hold min/max values).

File: src/hotspot/share/opto/c2_globals.hpp [hunk 8]
```diff
@@ -663,6 +663,16 @@
   develop(bool, VerifyAliases, false,                                       \
           "perform extra checks on the results of alias analysis")          \
                                                                             \
+  develop(uint, VerifyConstraintCasts, 0,                                   \
```
Comment (iwanowww) [inline src/hotspot/share/opto/c2_globals.hpp:666]: Any downsides in making the flag diagnostic? It'll make it available in product builds.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 9]
```diff
@@ -836,6 +836,52 @@ void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register reg_rax,
...
+  jccb(Assembler::less, fail);
+  cmpl(dst, type->_hi);
+  jccb(Assembler::lessEqual, succeed);
+  bind(fail);
+  movl(rax, dst);
+  movl(rcx, type->_lo);
+  movl(rdx, type->_hi);
+  hlt(); // hlt so we have the stack trace
```
Comment (iwanowww) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:852]: That's interesting. Sounds like a problem in `NativeStackPrinter::print_stack()`.

Speaking of debugging output, a call into a local helper function (encapsulating pretty printing logic) followed by a hlt call will do the job. But, considering the usages are in-line and quite common I suggest to make it conditional (guarded by a flag). It is possible to recover all 3 values from generated code if needed and turn on error reporting (specify the diagnostic flag) when reproducing failures.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 10]
```diff
@@ -7011,6 +7026,19 @@ instruct castLL(rRegL dst)
   ins_pipe(empty);
 %}
 
+instruct castLL_checked(rRegL dst, rRegL tmp, rFlagsReg cr)
```
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86_64.ad:7029]: Optimization idea: considering the range is statically known, `tmp` is not needed when range boundaries fit into signed int. Worth adding an extra AD instruction to benefit from that.

**Review by iwanowww — COMMENTED** [review]
Very nice!

**Comment by vnkozlov** [comment]
> @vnkozlov I don't have an AArch64 machine so I feel less confident writing one. We can add an AArch64 implementation later, though. What do you think?

Okay, later is fine.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 6 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: I have renamed it to `val`

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 9 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: I have made a helper function that will print the error message together with the parameters, the stack printing is still problematic, though.

File: src/hotspot/share/opto/c2_globals.hpp (hunk 8 above)
Comment (merykitty) [reply src/hotspot/share/opto/c2_globals.hpp]: That's a great idea, done it.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 11]
```diff
@@ -829,6 +829,84 @@ void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register reg_rax,
...
+    }
+    cmpl(val, type->_hi);
+    jccb(Assembler::lessEqual, succeed);
+  }
+
+  bind(fail);
+  movl(c_rarg0, val);
+  movl(c_rarg1, type->_lo);
+  movl(c_rarg2, type->_hi);
+  call(RuntimeAddress(CAST_FROM_FN_PTR(address, abort_checked_cast_int)));
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: I believe C2 stacks are 16-byte aligned so we should not need to manually do the alignment?

File: src/hotspot/cpu/x86/x86_64.ad [hunk 12]
```diff
@@ -7625,6 +7640,36 @@ instruct castLL(rRegL dst)
   ins_pipe(empty);
 %}
 
+instruct castLL_checked_L32(rRegL dst, rFlagsReg cr)
+%{
+  predicate(VerifyConstraintCasts > 0 &&
+            Assembler::is_simm32(static_cast<const CastLLNode*>(n)->type()->is_long()->_lo) &&
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/x86_64.ad]: Done, I added it to `ad_x86.hpp`

**Comment by merykitty** [comment]
@iwanowww Thanks for the reviews, I hope I have addressed all of your comments.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp [hunk 13]
```diff
@@ -44,6 +44,9 @@
                              Register t, Register thread);
   void fast_unlock_lightweight(Register obj, Register reg_rax, Register t, Register thread);
 
+  void checked_cast_int(const TypeInt* type, Register val);
```
[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp]: > Expanding the Type before passing into the function would be really bad after https://github.com/openjdk/jdk/pull/17508, though.

Can you elaborate, please? Are you saying that after #17508 there'll be more invariants to verify, so better to pass all the data encapsulated in a `Type`?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: Indeed. Still puzzled why stack traces are omitted. Do you have a reproducer to try?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 14]
```diff
@@ -829,6 +829,84 @@ void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register reg_rax,
...
+  hlt();
+  bind(succeed);
+  BLOCK_COMMENT("} // CastII");
+}
+
+static void abort_checked_cast_long(jlong val, jlong lo, jlong hi) {
+  fatal("Invalid CastLL, val: %lld, lo: %lld, hi: %lld", (long long)val, (long long)lo, (long long)hi);
```
[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: Thanks. Primarily, I had ideal node id in mind, but mach node info is useful as well.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 15]
```diff
@@ -422,6 +422,17 @@ source_hpp %{
 
 #include "peephole_x86_64.hpp"
 
+bool castLL_is_imm32(const Node* n);
+
+%}
+
+source %{
+
+bool castLL_is_imm32(const Node* n) {
```
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86_64.ad:431]: Please, assert that n is CastLL.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by iwanowww** [comment]
Speaking of bug synopsis, can you make it a bit more concrete and succinct?

How about "C2: Verify CastII/CastLL bounds at runtime"?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: You can just run `compiler/arraycopy/TestArrayCopyConjoint.java` with `-XX:+StressGCM -XX:VerifyConstraintCasts=2`.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp]: Yes that's what I mean, expanding all of them to feed into this method seems inelegant.

**Comment by merykitty** [comment]
@iwanowww Thanks a lot for the reviews, I have updated according to your suggestions.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp]: Ok, it makes perfect sense.

**Comment by iwanowww** [comment]
FTR here's AArch64 support:
  https://github.com/openjdk/jdk/commit/7ed34d09560d9db79183a80df379f3003f79bb1b

Feel free to incorporate it in this PR or I'll upstream it separately..

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: Mystery solved: it is the consequence of the fact that JVM doesn't preserve frame pointer in generated code. With `-XX:+PreserveFramePointer` everything works fine.

[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: FTR with the following patch I see stack traces both w/ and w/o `-XX:+PreserveFramePointer`:
https://github.com/iwanowww/jdk/commit/bbcd8d4dc7ab24152776a3e937ebd27367cf7331

I'll file a bug to fix the underlying issue.

**Comment by merykitty** [comment]
@iwanowww Thanks a lot for your help, I have incorporated your patches.

/contributor add vlivanov

**Review by iwanowww — APPROVED** [review]
Looks good. Thanks.

File: test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8346836
+ * @requires vm.debug == true & vm.flavor == "server"
+ * @summary Run with -Xcomp to test -XX:+StressGCM -XX:VerifyConstraintCasts=1 in debug builds.
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java:28]: Nit: you are also running it with `-XX:VerifyConstraintCasts=1`. I would just keep the summary more generic.
```suggestion
 * @summary Empty main program to run with flag VerifyConstraintCasts.
```

**Review by eme64 — APPROVED** [review]
Wow, this looks even much better with the improved printing on failure now!

Just out of curiosity: Is the whole `reconstruct_frame_pointer` mechanism general enough so that we could use it in other places as well? It is not super important to me any more, but I've wanted to have something like this for `VerifyAlignVector` already :)

File: test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java (hunk 16 above)
Comment (merykitty) [reply test/hotspot/jtreg/compiler/c2/TestVerifyConstraintCasts.java]: Thanks, I have done so

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp [hunk 17]
```diff
@@ -2743,3 +2743,107 @@ bool C2_MacroAssembler::in_scratch_emit_size() {
...
+    return;
+  }
+  BLOCK_COMMENT("verify_int_in_range {");
+  Label L_success, L_failure;
+
+  jint lo = t->_lo;
+  jint hi = t->_hi;
+
+  if (lo != min_jint && hi != max_jint) {
+    subsw(rtmp, rval, lo);
```
[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:2763]: Yes, you are right. Completely forgot that there's only 12 bits available for the immediate (`Assembler::operand_valid_for_add_sub_immediate()`).

[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:2763]: Completely forgot what I was thinking about when writing that code :-)

**Comment by merykitty** [comment]
@iwanowww @eme64 Thanks a lot for your reviews!
/integrate

**Comment by merykitty** [comment]
> IMO it is general enough, but I haven't found a good place to put it. Ideally, all runtime/debug calls should keep frame pointer valid for diagnostic purposes.

Tbh I don't understand ... How does the VM normally walk the stack when we crash in a compiled method? I thought that it has to know the frame size of a compiled method? Why do we need to manually reconstruct the frame pointer?

**Comment by iwanowww** [comment]
JVM knows how to unwind the stack when crash happens in compiled code (compiled frame on top). When native frame is on top, it relies on platform ABI, so fails to unwind the stack at the border of native and compiled frames because compiled code doesn't follow platform ABI conventions.
